web/modules/custom/dept_migrate/src/MigrateSupport.php

```diff
@@ -0,0 +1,269 @@
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   */
+  public function syncDomainsToGroups(EntityInterface $entity) {
+    if ($entity instanceof NodeInterface) {
+      $this->syncNodeDomainsToGroups($entity);
+    }
+    elseif ($entity instanceof UserInterface) {
+      $this->syncUserDomainsToGroups($entity);
+    }
+    else {
+      throw new \Exception('Unhandled entity type for ' . $entity->getEntityTypeId());
+    }
+  }
+
+  /**
+   * @param string $domain_machine_name
+   *   The machine name of the D7 domain.
+   * @return int
+   *   The group id (GID) of the group entity.
+   */
+  public function domainToGroupId(string $domain_machine_name) {
+    $group_id = 0;
+
+    switch ($domain_machine_name) {
+      case 'newnigov':
+      case 'del':
+      case 'dcal':
+      case 'doe':
+        $group_id = 1;
+        break;
+
+      case 'dfp':
+        $group_id = 2;
+        break;
+
+      case 'daera':
+        $group_id = 3;
+        break;
+
+      case 'communities':
+        $group_id = 4;
+        break;
+
+      case 'economy':
+        $group_id = 5;
+        break;
+
+      case 'education':
+        $group_id = 6;
+        break;
+
+      case 'execoffice':
+        $group_id = 7;
+        break;
+
+      case 'health':
+        $group_id = 8;
+        break;
+
+      case 'infrastructure':
+        $group_id = 9;
+        break;
+
+      case 'justice':
+        $group_id = 10;
+        break;
+    }
+
+    return $group_id;
+  }
+
+  /**
+   * @param \Drupal\node\NodeInterface $node
+   *   The node object.
+   */
+  protected function syncNodeDomainsToGroups(NodeInterface $node) {
+    // Fetch the map/details of the D7 entity.
+    $d7_entity = reset(\Drupal::service('dept_migrate.migrate_uuid_lookup_manager')->lookupByDestinationNodeIds([$node->id()]));
+
+    if (!empty($d7_entity['domains'])) {
+      $plugin_id = 'group_node-' . $node->bundle();
+
+      $group_content_data = $this->dbconn->query("
+          SELECT *
+          FROM {group_content_field_data}
+          WHERE entity_id = :id
+          AND type = :type", [
+            ':id' => $node->id(),
+            ':type' => 'department_site-' . $plugin_id,
+          ])->fetchAllAssoc('gid');
+
+      $node_group_ids = [];
+      foreach ($group_content_data as $gc_id => $row) {
+        $node_group_ids[] = $row->gid;
+      }
+
+      // Compare current groups to current domains, if they differ then
+      // update group content entity values.
+      $d7_mapped_group_ids = [];
+
+      foreach ($d7_entity['domains'] as $domain_machine_name) {
+        $d7_mapped_group_ids[] = $this->domainToGroupId($domain_machine_name);
+      }
+
+      $d7_mapped_group_ids = array_unique($d7_mapped_group_ids);
+
+      if (empty(array_diff($d7_mapped_group_ids, $node_group_ids))) {
+        // No changes, stop and return here.
+        return;
```

Comment:
Bailing out as early as possible here as we don't want to be deleting/re-adding the same thing each time, only if there are differences between D9 group content and D7 domain assignments.